### ISSUE TITLE
Add json output to tap

### ISSUE THREAD

**Issue by grampelberg** [linkerd/linkerd2#3390]
### What problem are you trying to solve?

It is a little tough to filter/read the current tap output. As headers are being added to tap, the output is starting to get difficult to consume. Take a peek at #3262 for an example. It would be nice to have some more machine readable output that can be sliced and diced with tools such as `jq`.

### How should the problem be solved?

```bash
$ linkerd -n linkerd tap deploy/linkerd-web -o json | jq .
{
    "source": {
        "ip": "10.4.1.42",
        "port": 55370
    },
    "source_meta": {
        "kind": "deployment",
        "namespace": "linkerd",
        "name": "linkerd-prometheus"
    },
    "destination": {
        "ip": "10.4.1.41",
        "port": 9994
    },
    "destination_meta": {
        "kind": "deployment",
        "namespace": "linkerd",
        "name": "linkerd-web"
    },
    "route_meta": {
        "route": "/api/"
    },
    "proxy_direction": "inbound",
    "http": {
        "request_init": {
            "id": {
                "base": 1234,
                "stream": 5678
            },
            "method": {
                "registered": "GET"
            },
            "scheme": {
                "registered": "HTTP"
            },
            "authority": "linkerd-web.linkerd.svc.cluster.local:9994",
            "path": "/api/foobar"
        }
    }
}
{
    "source": {
        "ip": "10.4.1.42",
        "port": 55370
    },
    "source_meta": {
        "kind": "deployment",
        "namespace": "linkerd",
        "name": "linkerd-prometheus"
    },
    "destination": {
        "ip": "10.4.1.41",
        "port": 9994
    },
    "destination_meta": {
        "kind": "deployment",
        "namespace": "linkerd",
        "name": "linkerd-web"
    },
    "route_meta": {
        "route": "/api/"
    },
    "proxy_direction": "inbound",
    "http": {
        "response_init": {
            "id": {
                "base": 1234,
                "stream": 5678
            },
            "since_request_init": {
                "seconds": 0,
                "nanos": 1000
            },
            "http_status": 200
        }
    }
}
{
    "source": {
        "ip": "10.4.1.42",
        "port": 55370
    },
    "source_meta": {
        "kind": "deployment",
        "namespace": "linkerd",
        "name": "linkerd-prometheus"
    },
    "destination": {
        "ip": "10.4.1.41",
        "port": 9994
    },
    "destination_meta": {
        "kind": "deployment",
        "namespace": "linkerd",
        "name": "linkerd-web"
    },
    "route_meta": {
        "route": "/api/"
    },
    "proxy_direction": "inbound",
    "http": {
        "response_end": {
            "id": {
                "base": 1234,
                "stream": 5678
            },
            "since_request_init": {
                "seconds": 0,
                "nanos": 1000
            },
            "since_request_init": {
                "seconds": 0,
                "nanos": 2000
            },
            "response_bytes": 1024,
            "eos": {
                "reset_error_code": 0
            }
        }
    }
}
```

Note: this is a JSON dump of the existing tap protobuf.

**Comment by olix0r** [comment]
I'd omit empty `headers` (especially in trailers) as that will be the norm there.

I feel strongly that response-length have a numerical value in json output.  Similarly `duration` should become `duration_us: 72`, etc.

In general, I'd preer more structure around `meta`. Some of these things are just `labels` (src_* dst_, others feel like they need not be nested under a `meta` at all (like id, proxy).


`:status: "200"`  should be numeric, and it's weird to encode the psuedoheader prefix when we're not talking about actual headers (which i believe _should_ include psuedoheaders).

**Comment by grampelberg** [comment]
> I'd omit empty headers (especially in trailers) as that will be the norm there.

I did it this way originally, reconsidered to make parsing/querying a little bit easier. I'm not super passionate about it, but I'd hate to have keys disappearing. 

> In general, I'd preer more structure around meta. Some of these things are just labels (src_* dst_, others feel like they need not be nested under a meta at all (like id, proxy).

I really like this idea, here's the tradeoff though. With this structure, you've got a 1:1 mapping to the normal output. If we move things around and add some more structure, it feels like we'll need to update the text output to match up as well. I think that'll be okay, just requires extra work.

**Comment by olix0r** [comment]
@grampelberg i wonder if we can simply render the protobuf as json? https://github.com/linkerd/linkerd2/blob/master/proto/public.proto#L183

How ugly is that?

**Comment by grampelberg** [comment]
```json
{
    "source": {
        "ip": "10.4.1.42",
        "port": 55370
    },
    "source_meta": {
        "kind": "deployment",
        "namespace": "linkerd",
        "name": "linkerd-prometheus"
    },
    "destination": {
        "ip": "10.4.1.41",
        "port": 9994
    },
    "destination_meta": {
        "kind": "deployment",
        "namespace": "linkerd",
        "name": "linkerd-web"
    },
    "route_meta": {
        "route": "/api/"
    },
    "proxy_direction": "inbound",
    "http": {
        "request_init": {
            "id": {
                "base": 1234,
                "stream": 5678
            },
            "method": {
                "registered": "GET"
            },
            "scheme": {
                "registered": "HTTP"
            },
            "authority": "linkerd-web.linkerd.svc.cluster.local:9994",
            "path": "/api/foobar"
        }
    }
}
{
    "source": {
        "ip": "10.4.1.42",
        "port": 55370
    },
    "source_meta": {
        "kind": "deployment",
        "namespace": "linkerd",
        "name": "linkerd-prometheus"
    },
    "destination": {
        "ip": "10.4.1.41",
        "port": 9994
    },
    "destination_meta": {
        "kind": "deployment",
        "namespace": "linkerd",
        "name": "linkerd-web"
    },
    "route_meta": {
        "route": "/api/"
    },
    "proxy_direction": "inbound",
    "http": {
        "response_init": {
            "id": {
                "base": 1234,
                "stream": 5678
            },
            "since_request_init": {
                "seconds": 0,
                "nanos": 1000
            },
            "http_status": 200
        }
    }
}
{
    "source": {
        "ip": "10.4.1.42",
        "port": 55370
    },
    "source_meta": {
        "kind": "deployment",
        "namespace": "linkerd",
        "name": "linkerd-prometheus"
    },
    "destination": {
        "ip": "10.4.1.41",
        "port": 9994
    },
    "destination_meta": {
        "kind": "deployment",
        "namespace": "linkerd",
        "name": "linkerd-web"
    },
    "route_meta": {
        "route": "/api/"
    },
    "proxy_direction": "inbound",
    "http": {
        "response_end": {
            "id": {
                "base": 1234,
                "stream": 5678
            },
            "since_request_init": {
                "seconds": 0,
                "nanos": 1000
            },
            "since_request_init": {
                "seconds": 0,
                "nanos": 2000
            },
            "response_bytes": 1024,
            "eos": {
                "reset_error_code": 0
            }
        }
    }
}
```

**Comment by grampelberg** [comment]
That's pretty decent honestly. There a little changes I'd make, but nothing substantial. WDYT @olix0r?

**Comment by grampelberg** [comment]
I updated to remove the headers (so that this work can happen in parallel) and for the first version of this it should match the existing protobuf.
